compose/ui/ui/src/uikitMain/kotlin/androidx/compose/ui/window/SkikoUIViewDelegate.kt

```diff
@@ -0,0 +1,154 @@
+import androidx.compose.ui.interop.UIKitInteropTransaction
+import androidx.compose.ui.scene.ComposeScene
+import androidx.compose.ui.scene.ComposeScenePointer
+import kotlin.math.floor
+import kotlin.math.roundToLong
+import kotlinx.cinterop.CValue
+import kotlinx.cinterop.useContents
+import org.jetbrains.skia.Canvas
+import platform.CoreGraphics.CGPoint
+import platform.Foundation.NSTimeInterval
+import platform.UIKit.UIEvent
+import platform.UIKit.UITouch
+import platform.UIKit.UITouchPhase
+import platform.UIKit.UIView
+
+internal enum class UITouchesEventPhase {
+    BEGAN, MOVED, ENDED, CANCELLED
+}
+
+internal interface SkikoUIViewDelegate {
+    fun pointInside(point: CValue<CGPoint>, event: UIEvent?): Boolean
+    fun onTouchesEvent(view: UIView, event: UIEvent, phase: UITouchesEventPhase)
+    fun retrieveInteropTransaction(): UIKitInteropTransaction
+    fun render(canvas: Canvas, targetTimestamp: NSTimeInterval)
+    var metalOffset: Offset
+}
+
+internal class SkikoUIViewDelegateImpl(
+    val sceneProvider: () -> ComposeScene,
+    val interopContext: UIKitInteropContext,
+    val densityProvider: DensityProvider,
+) : SkikoUIViewDelegate {
+    val scene get() = sceneProvider()
+    val density get() = densityProvider()
+    override var metalOffset: Offset = Offset.Zero
+
+    @Suppress("DEPRECATION")
+    override fun pointInside(point: CValue<CGPoint>, event: UIEvent?): Boolean =
+        point.useContents {
+            val position = Offset(
+                (x * density.density).toFloat(),
+                (y * density.density).toFloat()
+            )
+
+            !scene.hitTestInteropView(position)
+        }
+
+    override fun onTouchesEvent(view: UIView, event: UIEvent, phase: UITouchesEventPhase) {
+        scene.sendPointerEvent(
+            eventType = phase.toPointerEventType(),
+            pointers = event.touchesForView(view)?.map {
+                val touch = it as UITouch
+                val id = touch.hashCode().toLong()
+                val position = touch.offsetInView(view, density.density)
+                ComposeScenePointer(
+                    id = PointerId(id),
+                    position = position,
+                    pressed = touch.isPressed,
+                    type = PointerType.Touch,
+                    pressure = touch.force.toFloat(),
+                    historical = event.historicalChangesForTouch(touch, view, density.density)
+                )
+            } ?: emptyList(),
+            timeMillis = (event.timestamp * 1e3).toLong(),
+            nativeEvent = event
+        )
+    }
+
+    override fun retrieveInteropTransaction(): UIKitInteropTransaction =
+        interopContext.retrieve()
+
+    override fun render(canvas: Canvas, targetTimestamp: NSTimeInterval) {
+        // The calculation is split in two instead of
+        // `(targetTimestamp * 1e9).toLong()`
+        // to avoid losing precision for fractional part
```

Comment:
Please extract such things to separate functions. It should not be right in the `render` function